sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetReadState.java

```diff
@@ -33,31 +58,104 @@
   /** The remaining number of values to read in the current batch */
   int valuesToReadInBatch;
 
-  ParquetReadState(int maxDefinitionLevel) {
+  ParquetReadState(int maxDefinitionLevel, PrimitiveIterator.OfLong rowIndexes) {
     this.maxDefinitionLevel = maxDefinitionLevel;
+    this.rowRanges = rowIndexes == null ? null : constructRanges(rowIndexes);
```

Comment:
If we move this `rowIndexes == null ? null ...` part into `constructRanges` method, that will be better because it will protect both this constructor and `constructRanges` at the same time. For now, `constructRanges` seems to have the assumption that the argument is not `null` but there is no assertion for that.

Reply:
Yes that's a fair point. Will do.